lib/web/ui/usercontext.go

```diff
@@ -1,43 +1,99 @@
 package ui
 
 import (
+	"github.com/gravitational/teleport/lib/defaults"
 	"github.com/gravitational/teleport/lib/services"
-
-	"github.com/gravitational/trace"
+	"github.com/gravitational/teleport/lib/utils"
 )
 
+type access struct {
+	List   bool `json:"list"`
+	Read   bool `json:"read"`
+	Edit   bool `json:"edit"`
+	Create bool `json:"create"`
+	Delete bool `json:"remove"`
+}
+
+type userACL struct {
+	// Sessions defines access to recorded sessions
+	Sessions access `json:"sessions"`
+	// AuthConnectors defines access to auth.connectors
+	AuthConnectors access `json:"authConnectors"`
+	// Roles defines access to roles
+	Roles access `json:"roles"`
+	// TrustedClusters defines access to trusted clusters
+	TrustedClusters access `json:"trustedClusters"`
+	// SSH defines access to servers
+	SSHLogins []string `json:"sshLogins"`
+}
+
 type userContext struct {
 	// Name is this user name
 	Name string `json:"userName"`
-	// Email is this user email
-	Email string `json:"userEmail"`
-	// ACL is this user access control list
-	ACL RoleAccess `json:"userAcl"`
+	// ACL contains user access control list
+	ACL userACL `json:"userAcl"`
 }
 
-// NewUserContext returns userContext
-func NewUserContext(user services.User, allRoles []services.Role) (*userContext, error) {
-	userRoles := user.GetRoles()
-	roleNamesMap := map[string]bool{}
-	for _, name := range userRoles {
-		roleNamesMap[name] = true
+func getLogins(roleSet services.RoleSet) []string {
+	allowed := []string{}
+	denied := []string{}
+	for _, role := range roleSet {
+		denied = append(role.GetLogins(services.Deny), denied...)
```

Comment:
These should be:

```
denied = append(denied, role.GetLogins(services.Deny)...)
allowed = append(allowed, role.GetLogins(services.Allow)...)
```

At the moment you resetting the the slice at every iteration of the loop.